Add repeats to Reactive Messaging Tests

abutch3r opened the pull request:
Add repeats for various MP Reactive Messaging 3.0 combinations to the existing Reactive Messaging tests

Clean out Kafka Topics between executions as this is the quickest means to get a clean environment

Fix config tests to accommodate changes in logging between various JEE and MP

Fix framework to support new versions of MP Config via the use of `Optional` 

Make JsonbTest consistent with the other tests in using the test servlet as source

Fixes #25390 
Some Fixes for #25424

abutch3r committed:
Add repeats to Reactive Messaging Tests

Resolve issues in tests retlating to executing repeats in tests as existing tests had no prior need for handling the sitation were they were being rerun.

Clean out Kafka Topics between executions as this is the quickest means to get a clean environment and prevent future test corrutpion

Fix config validation tests to accommodate changes in logging between various JEE and MP

Make JsonbTest consistent with the other tests in using the test servlet and add repeats for RM combinations

abutch3r commented:
#libby #build

abutch3r commented:
#libby

abutch3r committed:
Improve Kafka topics cleanup

Use unique topic names in tests were the topic has >1 partitions.

Where we have tests failing to complete deletion - related to being partitioned. create unique names for the incoming topics. Outgoing topics have not been an issue as they are currently all single partition, so delete successfully.

Due to config requirements, dynamic topic names within the message tests need to be optional as the classes exist in those. However, the partition tests are executed in such a way that means optional is not required

Use CDI and MP Config to inject the topic names that are created into the classes that use

remove some of the attempted addtional hardening techniques.

Rename cleanup method to better match its actual behavior

Block on kafka deletion as admin commands are by default asynchronous if the code does not handle the kafkafutures that are created. by using `.get` we force it to block on best effort. However deletion can complete once all topics and their partitions are marked as `Marked for Deletion`

abutch3r commented:
#libby #build

abutch3r commented:
#build

abutch3r commented:
#build

abutch3r commented:
#libby #build

abutch3r commented on dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/invalid/badconfig/KafkaBadConfigTest.java:
```diff
@@ -214,7 +236,12 @@ public void testBadConfigOutgoingRetry() throws Exception {
         // Check that the bad config error was emitted
         List<String> configErrorLines = server.findStringsInLogsUsingMark("CWMRX1008E:", server.getDefaultLogFile());
-        assertThat(configErrorLines, hasSize(1));
+        //Due to logging changes older versions of OL emit the targetted code less frequently
```
need to change comment to be about older feature sets not OL itself

abutch3r commented on dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/kafka/invalid/badconfig/KafkaBadConfigTest.java:
```diff
@@ -177,7 +194,12 @@ public void testBadConfigOutgoing() throws Exception {
         // Check that the bad config error was emitted
         List<String> configErrorLines = server.findStringsInLogsUsingMark("CWMRX1008E:", server.getDefaultLogFile());
-        assertThat(configErrorLines, hasSize(1));
+        //Due to logging changes older versions of OL emit the expected code less frequently
```
need to change comment to be about older feature sets not OL itself

Azquelt replied:
Which bit of code did we change that makes these messages get logged more times?

abutch3r replied:
No Open Liberty Reactive Messaging code has changed which is the root of the original exception, based on the final exception `org.jboss.weld.exceptions.DeploymentException`, I suspect that it is the CDI versions that cause the change as the throwable that is in the FFDC descirption has grabbed different elements of the original exception.

Reactive Messaging-1.0 with CDI-2.0
```
org.jboss.weld.exceptions.DeploymentException: org.apache.kafka.common.KafkaException: Failed to construct kafka consumer: org.apache.kafka.common.config.ConfigException: No resolvable bootstrap urls given in bootstrap.servers com.ibm.ws.container.service.state.internal.ApplicationStateManager
```

Reactive Messaging-3.0 with CDI-4.0
```
org.jboss.weld.exceptions.DeploymentException: CWMRX1007E: A Kafka incoming connector for the bad-config-input channel cannot be created. The error is org.apache.kafka.common.KafkaException: Failed to construct kafka consumer: org.apache.kafka.common.config.ConfigException: No resolvable bootstrap urls given in bootstrap.servers com.ibm.ws.container.service.state.internal.ApplicationStateManager
```

In the first instance it has grabbed the KafkaException, while what is thrown in the logs is a KafkaConnectorException which wraps the exact same KafkaException we get an ffdc for in the first set of logs.

abutch3r commented:
This branch needs to be rebased once https://github.com/OpenLiberty/open-liberty/pull/25660 has been merged as that is changing the underlying  repeat actions definitions

also need to reorder the repeats so that the newest is done in LITE mode and the oldest is last

Azquelt commented on dev/com.ibm.ws.microprofile.reactive.messaging_fat/fat/src/com/ibm/ws/microprofile/reactive/messaging/fat/suite/KafkaUtils.java:
```diff
@@ -51,4 +61,56 @@ private static void copyFileToServer(File file, LibertyServer server) throws Exc
+     * @throws ExecutionException
+     * @throws InterruptedException
+     */
+    public static void deleteKafkaTopics(ExtendedKafkaContainer container) throws IOException, ExecutionException, InterruptedException {
```
This method is a bit horrible. We don't have enough information in the container itself to create the admin client so we're taking a guess at which class it came from based on the listener scheme and fishing the relevant properties out of there which isn't going to be reliable if we were to add more configurations.

abutch3r replied:
Can move the credential values from the Sasl test class into ExtendedKafkaContainer so that if other tests in the future need authentication details then we can easily reuse them and its cleaner from the perspective of cleanup as it is now from a common source accessible to all our containers.

Would need to add more helpers for the various properties or make it easier to pull the various Jaas configs from the container if we want to start exposing more of the security details to any operation wanting to build a connection to kafka

abutch3r committed:
Update Repeat action identifiers

Create kafka AdminClients in the classes which define the kafka containers and pass those through to the deletion method so that the utils doesn't have to know about any security details that the defining class does.

Change how the war for Invalid Liberty login is loaded into the server to stop it reusing the same app for each repeat

Update connectionProperties definitions to produce Map<String, Object> so that they can be more widely used, such as helping create the AdminClients

abutch3r commented:
#build

abutch3r commented:
#libby

abutch3r commented:
#libby #build
